storage/innobase/log/log0recv.cc

```diff
@@ -1727,9 +1727,20 @@ static byte *recv_parse_or_apply_log_rec_body(
         if (page_no == 0 && !fsp_is_system_or_temp_tablespace(space_id) &&
             /* For cloned db header page has the encryption information. */
             !recv_sys->is_cloned_db) {
-          if (fil_tablespace_redo_encryption(ptr, end_ptr, space_id) == nullptr)
-            return (nullptr);
+          byte *ptr_copy = ptr;
+          ptr_copy += 2;  // skip offset
+          ulint len = mach_read_from_2(ptr_copy);
+          ptr_copy += 2;
+          if (end_ptr < ptr_copy + len) return NULL;
+          if (memcmp(ptr_copy, ENCRYPTION_KEY_MAGIC_V3,
+                     ENCRYPTION_MAGIC_SIZE) == 0) {
+            return (fil_tablespace_redo_encryption(ptr, end_ptr, space_id));
+          } else if (memcmp(ptr_copy, ENCRYPTION_KEY_MAGIC_PS_V2,
+                            ENCRYPTION_MAGIC_SIZE) == 0) {
+            break;
```

Comment:
this works? surprising... the v2 writer is only writing KERYING_ENCRYPTION_INFO_MAX_SIZE bytes.

mlog_write_string(page + offset, encrypt_info,
                    KERYING_ENCRYPTION_INFO_MAX_SIZE, mtr);

Also you need to handle V1 KERYING_ENCRYPTION_INFO_MAX_SIZE_V1 bytes.

Reply:
Also there should be ptr null check